File: leetcode/00020. Valid Paranthesis/Python/sol.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        match = {')': '(', ']': '[', '}': '{'}
        for c in s:
            if c in "([{":
                st.append(c)
            elif not st or st[-1] != match[c]:
                return False
            else: 
                st.pop()
        return not st

File: leetcode/00020. Valid Paranthesis/C++/sol.cpp
#include <stack>
#include <unordered_map>
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> st;
        std::unordered_map<char, char> match = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };
        for (char c : s) {
            if (c == '(' || c == '[' || c == '{') st.push(c);
            else if (st.empty() || st.top() != match[c]) return false;
            else st.pop();
        }
        return st.empty();
    }
};